From d5b5b27e2776f77ba5c5f614177559888b9b8861 Mon Sep 17 00:00:00 2001
From: Frank Wang <frank.wang@rock-chips.com>
Date: Thu, 26 Mar 2020 10:59:49 +0800
Subject: [PATCH] mailbox: rockchip: fix to null pointer dereference

Delivery message to upper only when mailbox client is existed.

Signed-off-by: Frank Wang <frank.wang@rock-chips.com>
Change-Id: I2fbfc3ecd110fe26d0ffed5ee90d1e4877450ef9
---
 drivers/mailbox/rockchip-mailbox.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/drivers/mailbox/rockchip-mailbox.c b/drivers/mailbox/rockchip-mailbox.c
index d1a629cc01e4..9251ec5291fd 100644
--- a/drivers/mailbox/rockchip-mailbox.c
+++ b/drivers/mailbox/rockchip-mailbox.c
@@ -123,11 +123,12 @@ static irqreturn_t rockchip_mbox_irq(int irq, void *dev_id)
 			msg.data = readl_relaxed(mb->mbox_base +
 						 MAILBOX_B2A_DAT(idx));
 
-			mbox_chan_received_data(&mb->mbox.chans[idx], &msg);
-
 			dev_dbg(mb->mbox.dev, "Chan[%d]: B2A message, cmd 0x%08x, data 0x%08x\n",
 				idx, msg.cmd, msg.data);
 
+			if (mb->mbox.chans[idx].cl)
+				mbox_chan_received_data(&mb->mbox.chans[idx],
+							&msg);
 			/* Clear mbox interrupt */
 			writel_relaxed(1U << idx,
 				       mb->mbox_base + MAILBOX_B2A_STATUS);
-- 
2.35.3

